### PR TITLE
Fixed #26355 -- Added API to register different expressions for operator/field combo.

### PR DESCRIPTION
**Pull request by mrzechonek**
The idea is to register expression class in `CombinedExpression` lookup
dict so that it can construct i.e. a `Func` when combining two sides with
Python's `+` operator.

### CODE REVIEW

**Comment by mrzechonek** [comment]
@mjtamlyn This is the code I've been asking about during DjangoCon sprints. Will do more work on that.

**Comment by mjtamlyn** [comment]
@jarshwah This is what I was mentioning to you on IRC. I'd also be very interesting in @akaariai's opinins.

**Comment by jarshwah** [comment]
Love it! I've been thinking about how to handle mixed combinable operators for awhile, and this looks like a really good way to go. Particularly fond of some of the stand-in comments and error messages :smile: 

My early thoughts were that fields might know how they connect with theirselves and others, by either providing a method that did operator resolution or by implementing `__radd` methods. Combinable would then inspect `output_field` on lhs and rhs and extract the appropriate connector from the field. Expressions aren't fields though, and fields are at a lower abstraction level and shouldn't really have knowledge of expressions. As your examples and the ones in the tickets show, combining with an infix operator isn't always the correct thing to do either, so injecting expressions into the resolution process seems like the right thing to do.

A few notes. 

- Do the left and right hand sides need to be strings/names? If we can get away with registering the field types themselves I think that'd be preferable.

```
@CombinedExpression.register_expression(ArrayField, CombinedExpression.ADD, ArrayField)
```

- I think I'd prefer `register_expression` to have a name that more closely resembles operator overloading. Something like `register_overload`, though that's a poor name itself.

- Shortcuts for registering combinations might be useful (or not after writing an example):

```
@CombinedExpression.register_expression(['ArrayField', None], CombinedExpression.ADD, ['ArrayField', None])

combos = itertools.product(lhs, connectors, rhs)
[('ArrayField', 'ADD', 'ArrayField'),
 ('ArrayField', 'ADD', None),
 (None, 'ADD', 'ArrayField'),
 (None, 'ADD', None)]
# (None, None) is unfortunate for this example.. which makes me reconsider the usefulness
```

**Comment by akaariai** [comment]
Big +1 to the idea. 

I have a feeling we could do something better for the nicely commented _lookup_expression method. The basic idea would be to create a dictionary which handles the mro issue internally. I'll try to come up with some code as soon as I have time.

**Comment by mrzechonek** [comment]
Agree. Names are temporary, was just trying out the idea.

I'll also take a shot at MRO lookup. One improvement would be to register a Field instead of None for catch-all.

As for types vs names, this is problematic for circular imports. Consider TemporalSubtraction. I couldn't find an easy way to move it, but maybe a separate file (operators.py, maybe?) would do the trick.

**Comment by charettes** [comment]
Great work here!

I remember discussing a way of exposing such an API with Josh at the Field level (here's the [commit](https://github.com/charettes/django/commit/cfa00ff2f64649a943fa75db9ff6ea62878073ef)) but just like him I think your approach makes more sense.

I think an `operators` module would make sense.

**Comment by akaariai** [comment]
See https://gist.github.com/akaariai/08952c3dc7c24e685ea670d45b5aec91 for an approach using the same rules than what Python uses for finding the operator.

Notably this results in registering operators as (klass, operator) -> Expression. That is, for example (DateField, SUB) -> TemporalSubtraction. And, according to these rules, F('date_field') - Value(1, output_field=IntegerField()) will be resolved as TemporalSubtraction. So, I'm not sure if we want to go by the Python rules after all.

Maybe we could have both, that is you can register both klass1 op klass2 style operators, and also "klass op" style rules, which use Python resolution rules? In any case we should make sure the rules are well defined, and of course that the rules are useful for real world usage.

**Comment by mrzechonek** [comment]
Well, Python doesn't do overloading at all, so I'd rather look at how SQL does it, instead of abusing the rules. We're building a query after all.

I kinda know the algorithm for overload resolution in c++, but it's crazy complex even without ambiguity detection. I'm worried about performance.

I think in this case it's just a case of a (polymorphic) double dispatch, so there should be something available already.

**Comment by akaariai** [comment]
In python you have well defined rules what happens when you calculate date + integer for example. So, if I'm not mistaken, python does have rules for polymorphic double dispatch for operators.

Another gist with a small addition: possibility to add klass1 op klass2 style operators, too. See https://gist.github.com/akaariai/c7bab4d4da56c772ed48329fdc328aa0

**Comment by mrzechonek** [comment]
As far as I understand the data model, `a <op> b`  is `a.__<op>__(b)` regardless of `b`'s type. This seems like a single dispatch to me... and also splitting hairs ;).

Regardless, since Python is all about `pip install the_feature` and "I hate code and I want it to have as little to do with our product as possible", what would be your opinion on using https://pypi.python.org/pypi/multipledispatch for this?

**Comment by akaariai** [comment]
Actually, if b is subclass of a, then a op b is b.op(a).

As you said, this isn't that important. As long as the rules are easy to understand and covering enough, then I'm fine. I'm not sure the Python rules are good enough.

If multipledispatch has sane rules, I'm ok with it.

**Comment by mrzechonek** [comment]
I've updated the code to use bundled `multipledispatch`. This allowed me to remove funny comments, friendly exceptions and enabled registration by type, not a string.

If this looks like a step in right direction (particularly the bundling, I asked on the mailing list but there was no response really), I'll fix isort/flake errors and start expanding the test suite.

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -394,14 +395,26 @@ def as_sql(self, compiler, connection):
...
             return DurationExpression(self.lhs, self.connector, self.rhs).as_sql(compiler, connection)
-        if (lhs_output and rhs_output and self.connector == self.SUB and
-            lhs_output.get_internal_type() in {'DateField', 'DateTimeField', 'TimeField'} and
-                lhs_output.get_internal_type() == lhs_output.get_internal_type()):
-            return TemporalSubtraction(self.lhs, self.rhs).as_sql(compiler, connection)
+
+        if lhs_output or rhs_output:
+            dispatcher = self.connector_dispatchers.get(self.connector)
+            try:
+                combined_expression = dispatcher(lhs_output, rhs_output) if dispatcher else CombinedExpression
```
Comment (akaariai) [inline django/db/models/expressions.py]: Should we get rid of defaulting to CombinedExpression and instead use the registration machinery for CombinedExpression, too?

File: django/db/models/operators.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+from django.db.models.fields import (
+    DateField, DateTimeField, DurationField, TimeField
+)
+
+__all__ = ()
+
+
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)
+@CombinedExpression.dispatch(CombinedExpression.SUB, TimeField, TimeField)
```
Comment (akaariai) [inline django/db/models/operators.py]: DateTimeField is a subclass of DateField, so no need for the second case?

File: django/contrib/postgres/fields/array.py [hunk 3]
```diff
@@ -288,3 +289,13 @@ def __init__(self, start, end):
 
     def __call__(self, *args, **kwargs):
         return SliceTransform(self.start, self.end, *args, **kwargs)
+
+
+@CombinedExpression.dispatch(CombinedExpression.ADD, ArrayField, None)
+@CombinedExpression.dispatch(CombinedExpression.ADD, None, ArrayField)
+@CombinedExpression.dispatch(CombinedExpression.ADD, ArrayField, ArrayField)
```
Comment (akaariai) [inline django/contrib/postgres/fields/array.py]: Do we need the third case, isn't ArrayField, None enough to cover ArrayField, ArrayField, too.

Field would read much better than None for the catch-all case.

What happens if you have also a rule CharField, Field -> StringConcat, and you run a query with F('char_field') + F('array_field')?

**Comment by akaariai** [comment]
Apart of some minor issues, this looks really nice. Great work!

File: django/db/models/operators.py (hunk 2 above)
Comment (charettes) [inline django/db/models/operators.py]: This could end up being problematic as only subtraction of same temporal types is actually supported.

Comment (akaariai) [inline django/db/models/operators.py]: I believe you can already do DateField - DateTimeField with these rules.

Comment (charettes) [inline django/db/models/operators.py]: The thing is this was left unsupported on purpose. See #5997 lasts comments.

Comment (akaariai) [inline django/db/models/operators.py]: Sorry, I meant that .distpatch(CombinedExpression.SUB, DateField, DateField) will allow DateTimeField - DateField.

I guess we can just fail during execution if one of the fields is DateTimeField and the other a DateField.

File: django/contrib/postgres/fields/array.py (hunk 3 above)
Comment (mrzechonek) [inline django/contrib/postgres/fields/array.py]:
```
Ambiguities exist in dispatched function +

The following signatures may result in ambiguous behavior:
	[CharField, Field], [Field, ArrayField]
	[ArrayField, Field], [Field, CharField]


Consider making the following additions:

@dispatch(CharField, ArrayField)
def +(...)

@dispatch(ArrayField, CharField)
def +(...)
```
This is raised from `multipledispatch` on registration (!), not during query execution. I think it's a Good Thing (TM).

File: django/db/models/expressions.py (hunk 1 above)
Comment (mrzechonek) [inline django/db/models/expressions.py]: The fallback is meant for unregistered operators.

Speaking of which: I think I might be able to get rid of `connection.ops.combine_expression` call in favor of dispatching expression suitable for the particular database backend. This will be a bit more invasive, but I think also cleaner in the long run. What do you think?

File: django/db/models/operators.py (hunk 2 above)
Comment (mrzechonek) [inline django/db/models/operators.py]: Good catch, I'll make sure to check this.

File: django/contrib/postgres/fields/array.py (hunk 3 above)
Comment (mrzechonek) [inline django/contrib/postgres/fields/array.py]: Also, `ArrayField + ArrayField` is ambiguous wrt `ArrayField + Field`, so we need all 3 dispatches. 
There might be a way to avoid raising warnings if all dispatches point to the same implementation. I'll give it a shot.

File: django/db/models/expressions.py (hunk 1 above)
Comment (jarshwah) [inline django/db/models/expressions.py]: If you want to give that a go then I'd recommend doing so in a separate commit so we can see exactly the changes required (and not clutter up the feature implementation).

**Comment by auvipy** [comment]
@mrzechonek can we expect this feature on 1.10 release cycle?

**Comment by mrzechonek** [comment]
@auvipy given the timeline, I don't think so, sorry.

I'm having issues with combinatorial explosion, still trying to trim the rules down without causing conflicts/ambiguities.

File: django/db/models/operators.py (hunk 2 above)
Comment (patrys) [inline django/db/models/operators.py]: Resurrecting this half a year later: IMHO most subclasses should work by default except a few known cases. For example I'd expect MyCustomDateField - DateField to just work but expect subclasses of DateTimeField to fail when combined with subclasses of DateField.

Stopping on first match is actually enough to handle this now but may be overly verbose:

```python
def handle(...):
    pass

def reject(...):
    raise NotImplementedError(...)

CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)(handle)

CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateField)(reject)
CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateTimeField)(reject)

CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)(handle)
```

Possible solution would be to explicitly mark unhandled cases (keep in mind this would be mostly used internally by Django's built in set of fields):

```python
@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateField, unhandled=True)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateTimeField, unhandled=True)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)
def handle(...):
    pass
```

An easier fix would be to introduce a base class for the common part of DateField and DateTimeField (and other cases where it's convenient to share code but the types don't really make sense when combined) but it'd likely break backwards compatibility.

**Comment by mrzechonek** [comment]
I've just updated the implementation to resolve the issues with ambiguous dispatches. Tests seem to work, including the original feature request.

Would appreciate another round of reviews!

**Comment by felixxm** [comment]
Please check [isort](http://djangoci.com/job/isort/12854/console) and [flake](http://djangoci.com/job/flake8/12863/console) details.

File: django/db/models/expressions.py [hunk 4]
```diff
@@ -372,77 +376,88 @@ def as_sql(self, compiler, connection):
...
+                    # avoid recursion in case CombinedExpression maps to another
+                    # CombinedExpression, but with a different connector
+                    if type(self) != klass:
+                        try:
+                            output_field = self._output_field_or_none
+                        except FieldError:
+                            output_field = None
+
+                        expression = klass(self.lhs, self.connector, self.rhs, output_field=output_field)
+                        print(expression)
```
Comment (francoisfreitag) [inline django/db/models/expressions.py]: Looks like a forgotten print statement.

File: django/db/models/expressions.py [hunk 5]
```diff
@@ -596,6 +611,9 @@ def get_group_by_cols(self):
 
 
 class DurationValue(Value):
+    def __init__(self, value, output_field=None):
+        super(DurationValue, self).__init__(value, output_field=fields.DurationField())
```
Comment (francoisfreitag) [inline django/db/models/expressions.py]: What if `output_field` is given to the constructor?

File: django/db/models/expressions.py (hunk 4 above)
Comment (mrzechonek) [inline django/db/models/expressions.py]: Removed.

**Comment by mrzechonek** [comment]
About requiring explicit `output_field` on `Value` expressions, seems like SQLAlchemy does the same thing: http://docs.sqlalchemy.org/en/latest/core/sqlelement.html#sqlalchemy.sql.expression.literal

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/db/models/operators.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+__all__ = ()
+
+
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, TimeField, TimeField, commutative=False)
+class TemporalSubtraction(CombinedExpression):
+    def __init__(self, lhs, connector, rhs, output_field):
+        super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs, output_field=DurationField())
```
Comment (francoisfreitag) [inline django/db/models/operators.py]: I fail to understand why the `output_field` passed is replaced by a `DurationField()`? I would have expected:
```python
super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs, output_field=output_field or DurationField())
```

**Comment by mrzechonek** [comment]
This class was moved from expressions.py, where it was like that. I haven't touched it.

I suspect this is to override field type in backends that resolve datetime subtractions to integer fileds (number of seconds IIRC)

File: django/db/models/expressions.py [hunk 7]
```diff
@@ -372,77 +376,87 @@ def as_sql(self, compiler, connection):
...
+                    # avoid recursion in case CombinedExpression maps to another
+                    # CombinedExpression, but with a different connector
+                    if type(self) != klass:
+                        try:
+                            output_field = self._output_field_or_none
+                        except FieldError:
+                            output_field = None
+
+                        expression = klass(self.lhs, self.connector, self.rhs, output_field=output_field)
+                        return compiler.compile(expression)
```
Comment (charettes) [inline django/db/models/expressions.py]: Keep the `try` body to a minimum by only including the code that could raise the expected `NotImplementedError` exception. Use an `else` branch if necessary.

**Comment by charettes** [comment]
I wonder if could depend on `multipledispatch` like we do with `pytz` instead of vendoring it now that we consider Python packaging _good_.

**Comment by timgraham** [comment]
Consensus will be required on the django-developers mailing list about whether to vendor or depend on multidispatch.

Documentation for the new API seem needed?

**Comment by jarshwah** [comment]
Post for vendor vs dependency: https://groups.google.com/forum/#!topic/django-developers/IKCHOrlxr5w

@mrzechonek are you planning to address comments above before the feature freeze of 1.11 (January 16, 2017)? If not, maybe someone else could polish the patch for merge. Keep the vendoring of multipledispatch for now. If the ML agrees to add as a dependency we can make the change then.

**Comment by mrzechonek** [comment]
@jarshwah I'll do my best to meet the deadline

**Comment by mrzechonek** [comment]
@timgraham docs would indeed be useful, would it be OK to create a new file called `operators.txt` in https://github.com/django/django/tree/master/docs/ref/models?

File: django/db/models/expressions.py (hunk 7 above)
Comment (mrzechonek) [inline django/db/models/expressions.py]: :+1:

File: django/db/models/operators.py (hunk 6 above)
Comment (mrzechonek) [inline django/db/models/operators.py]: :+1:

Comment (mrzechonek) [inline django/db/models/operators.py]: @charettes unfortunately I can't instantiate a `Field` at the module level (I get `ImproperlyConfigured` as visible in Jenkins), so I've changed the invocation to `output_field=output_field or DurationField()`

**Comment by timgraham** [comment]
About the docs location, feel free to do whatever you think is best. The main work is the content. Once everyone sees that, where to put it can be easily changed if needed.

**Comment by jarshwah** [comment]
The mailing list thread has been somewhat quiet, but we've had 5 for adding a dep and none for vendoring. A question that was brought up was what version specifier should we use. Aymeric put forward that dependency versions are a project concern, and unless we have any firm requirements, we shouldn't have a version requirement at all.

**Comment by mrzechonek** [comment]
Ok, so I guess I should revert the vendoring and add a dependency in `setup.py`?

**Comment by timgraham** [comment]
Yes, also please add to `tests/requirements/base.txt` (and to the 1.11 release notes).

**Comment by timgraham** [comment]
Will you try to complete the documentation this week to make the 1.11 feature freeze?

**Comment by mrzechonek** [comment]
I'm working on that right now, will submit an update shortly.

By the way, I have no idea why Jenkins complains about flake8 and windows, I think I've fixed both issues...

**Comment by mrzechonek** [comment]
@timgraham I've added some docs, hope that's enough.

File: tests/model_fields/test_charfield.py [hunk 8]
```diff
@@ -32,6 +33,14 @@ def test_emoji(self):
         p.refresh_from_db()
         self.assertEqual(p.title, 'Smile 😀')
 
+    def test_add(self):
```
Comment (timgraham) [inline tests/model_fields/test_charfield.py]: This probably isn't the best place for the test -- it's testing changes in expressions, not in model fields. Try `tests/expressions` instead.

File: tests/model_fields/test_charfield.py [hunk 9]
```diff
@@ -32,6 +33,14 @@ def test_emoji(self):
...
+        instance = Post.objects.create(title='Test', body='Post')
+        Post.objects.update(title=F('title') + Value('_title', output_field=CharField()),
```
Comment (timgraham) [inline tests/model_fields/test_charfield.py]: Use hanging indent for cases like this:
```
Post.objects.update(
    title=F('title') + Value('_title', output_field=CharField()),
    ....
)
```

File: django/db/models/expressions.py [hunk 10]
```diff
@@ -33,6 +35,8 @@ class Combinable(object):
     BITLEFTSHIFT = '<<'
     BITRIGHTSHIFT = '>>'
 
+    ANY = None
```
Comment (timgraham) [inline django/db/models/expressions.py]: Does this deserve a comment? The purpose isn't obvious to me.

File: django/db/models/expressions.py [hunk 11]
```diff
@@ -380,77 +385,87 @@ def as_sql(self, compiler, connection):
             lhs_output = self.lhs.output_field
         except FieldError:
             lhs_output = None
+
```
Comment (timgraham) [inline django/db/models/expressions.py]: please revert an unrelated whitespace change

File: django/db/models/expressions.py [hunk 12]
```diff
@@ -380,77 +385,87 @@ def as_sql(self, compiler, connection):
...
         try:
             rhs_output = self.rhs.output_field
         except FieldError:
             rhs_output = None
+
+        # for databases that don't do temporal interval types, skip the
```
Comment (timgraham) [inline django/db/models/expressions.py]: use capitals to start a sentence please

File: django/db/models/expressions.py [hunk 13]
```diff
@@ -380,77 +385,87 @@ def as_sql(self, compiler, connection):
...
 
+    @classmethod
+    def dispatch(cls, connector, lhs, rhs, commutative=True):
+        dispatcher = cls.expression_dispatchers.setdefault(connector, multipledispatch.Dispatcher(connector))
 
-class TemporalSubtraction(CombinedExpression):
-    def __init__(self, lhs, rhs):
-        super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs, output_field=fields.DurationField())
+        def _inner(klass):
+            try:
```
Comment (timgraham) [inline django/db/models/expressions.py]: Is this try/except clause as minimal as can be? It's not obvious where the `TypeError` might come from. Some comments may help.

File: docs/ref/models/expressions.txt [hunk 14]
```diff
@@ -728,3 +728,47 @@ file (or package) that is imported from the top level ``__init__.py``.
 
 For user projects wishing to patch the backend that they're using, this code
 should live in an :meth:`AppConfig.ready()<django.apps.AppConfig.ready>` method.
+
+Using operators to combine expressions
+--------------------------------------
+
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: Add `.. class:: CombinedExpression(...)` as done elsewhere.

File: docs/ref/models/expressions.txt [hunk 15]
```diff
@@ -728,3 +728,47 @@ file (or package) that is imported from the top level ``__init__.py``.
...
+  >>> qs.values_list('full_name')
+  <QuerySet [(u'Stephen Keats',)]>
+
+This works by registering such combinations using ``CombinedExpression.dispatch``
+decorator, allowing Django to generate correct SQL equivalent for Python
+operators for a given operator and left- and right-hand side operands.
+
+Therefore, string concatenation is registered as follows::
+
+  from django.db.models.functions import ConcatPair
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: Add import for CombinedExpression

**Comment by mrzechonek** [comment]
Thanks for a review!

**Comment by mrzechonek** [comment]
Thanks for a review! I've taken care of the issues and split the PR into 2 commits, as suggested.

**Comment by mrzechonek** [comment]
@timgraham Any chances for a merge? :)

**Comment by auvipy** [comment]
@mrzechonek I can see a merge conflict

**Comment by mrzechonek** [comment]
@auvipy Got it, rebased.

File: django/db/models/operators.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+# dispatch expressions
+__all__ = ()
+
+
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, TimeField, TimeField, commutative=False)
+class TemporalSubtraction(CombinedExpression):
+    def __init__(self, lhs, connector, rhs, output_field=None):
+        super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs,
```
Comment (jarshwah) [inline django/db/models/operators.py]: I'm very much not a fan of this kind of indentation. I much prefer the form:

```
super(TemporalSubtraction, self).__init__(
    lhs, self.SUB, rhs, output_field=output_field or DurationField())
```

Comment (jarshwah) [inline django/db/models/operators.py]: There are other examples of this indentation style throughout, I won't comment on each one. Also, I realise my preference for style is kind of arbitrary, but it *is* mostly consistent through the code base already.

Comment (mrzechonek) [inline django/db/models/operators.py]: Not a problem. I've updated indentation in the whole patch.

**Comment by timgraham** [comment]
Sorry, this didn't make it to 1.11. I didn't get to a second detailed review and felt it might be better not to add another dependency days before the feature freeze (especially since this is the first version of Django with a dependency). Also, given some discussion on the mailing list about a possible alternative to multidispatch.

**Comment by mrzechonek** [comment]
No problem, you're the boss.

If there's anything else, I'm happy to keep this PR alive.

File: django/contrib/postgres/fields/array.py [hunk 17]
```diff
@@ -295,3 +296,11 @@ def __init__(self, start, end):
...
     def __call__(self, *args, **kwargs):
         return SliceTransform(self.start, self.end, *args, **kwargs)
+
+
+@CombinedExpression.dispatch(CombinedExpression.ADD, ArrayField, ArrayField)
+class ArrayCatPair(Func):
+    function = 'ARRAY_CAT'
+
+    def __init__(self, lhs, connector, rhs, output_field):
+        super(ArrayCatPair, self).__init__(lhs, rhs, output_field=output_field)
```
Comment (timgraham) [inline django/contrib/postgres/fields/array.py]: You can rebase your branch and target it for Django 2.0. Since master no longer supports Python 2, you can make a few updates such as using `super().`.

**Comment by mrzechonek** [comment]
Done.

**Comment by charettes** [comment]
Now that this targets 2.0, which is Python 3 only, I wonder if it could be better to use [overloading](https://pypi.python.org/pypi/overloading) instead as suggested by [Adam](https://groups.google.com/d/msg/django-developers/IKCHOrlxr5w/W13sNkOwEwAJ).

**Comment by mrzechonek** [comment]
From what I see in overloading.py, there is no public API to [register overloads by hand](
https://github.com/mrzechonek/django/blob/65fbecbe11bcb8e58f5e30fffbcf1eac29753624/django/db/models/expressions.py#L453), using manually constructed signature. It relies on [introspection of type hints](https://github.com/bintoro/overloading.py/blob/master/overloading.py#L199) instead.

**Comment by mrzechonek** [comment]
BTW I have a feeling that operator overloading would also be useful for [#25367](https://code.djangoproject.com/ticket/25367) and we should use the same library for both.

**Comment by jarshwah** [comment]
Regarding `overloading` and not being able to register by hand, that's pretty much the conclusion I came to when it was proposed. Unless someone has more to offer WRT overloading, are we ok to merge this?

File: django/db/models/operators.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+__all__ = ()
+
+
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField, commutative=False)
+@CombinedExpression.dispatch(CombinedExpression.SUB, TimeField, TimeField, commutative=False)
+class TemporalSubtraction(CombinedExpression):
+    def __init__(self, lhs, connector, rhs, output_field=None):
+        super().__init__(
+            lhs, self.SUB, rhs, output_field=output_field or DurationField())
```
Comment (timgraham) [inline django/db/models/operators.py]: The closing parenthesis in cases like this goes on the next line.

File: docs/releases/1.11.txt [hunk 19]
```diff
@@ -352,6 +352,11 @@ Models
 
 * Added :meth:`.QuerySet.union`, :meth:`~.QuerySet.intersection`, and
   :meth:`~.QuerySet.difference`.
+* Added support for operator overloading to create
```
Comment (timgraham) [inline docs/releases/1.11.txt]: Move to 2.0.

File: docs/ref/models/expressions.txt [hunk 20]
```diff
@@ -893,3 +893,53 @@ file (or package) that is imported from the top level ``__init__.py``.
...
 For user projects wishing to patch the backend that they're using, this code
 should live in an :meth:`AppConfig.ready()<django.apps.AppConfig.ready>` method.
+
+Using operators to combine expressions
+--------------------------------------
+
+.. versionadded:: 1.11
+
+.. currentmodule:: django.db.models.expressions
+.. class:: CombinedExpression(...)
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: `(...)` seems inconsistent

File: docs/releases/2.0.txt [hunk 21]
```diff
@@ -168,6 +168,12 @@ Migrations
 Models
 ~~~~~~
 
+* Added support for operator overloading to create
+  :class:`~django.db.models.expressions.CombinedExpression`. This means that ``F('text_field') +
+  Value('_suffix', output_field=CharField())`` now results in ``ConcatPair``
+  invocation instead of verbatim addition. Because of this, Django now
```
Comment (timgraham) [inline docs/releases/2.0.txt]: Maybe "``ConcatPair`` invocation instead of verbatim addition" could be explained with an example.

File: docs/ref/models/expressions.txt [hunk 22]
```diff
@@ -893,3 +893,53 @@ file (or package) that is imported from the top level ``__init__.py``.
 
 For user projects wishing to patch the backend that they're using, this code
 should live in an :meth:`AppConfig.ready()<django.apps.AppConfig.ready>` method.
+
+Using operators to combine expressions
+--------------------------------------
+
+.. versionadded:: 2.0
+
+.. currentmodule:: django.db.models.expressions
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: Should `CombinedExpression` have a convenience import in `django.db.models` like the other public expressions?

File: docs/ref/models/expressions.txt [hunk 23]
```diff
@@ -893,3 +893,53 @@ file (or package) that is imported from the top level ``__init__.py``.
...
+    >>> from django.db.models import F, Value, CharField
+    >>> qs = Author.objects.annotate(
+    ...    full_name=(
+    ...       F('first_name') +
+    ...       Value(' ', output_field=CharField()) +
+    ...       F('last_name')))
+    >>> qs.values_list('full_name')
+    <QuerySet [('Stephen Keats',)]>
+
+This works by registering such combinations using ``CombinedExpression.dispatch``
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: It might be useful to document: `.. method:: CombinedExpression.dispatch(...)`

File: docs/ref/models/expressions.txt [hunk 24]
```diff
@@ -893,3 +893,53 @@ file (or package) that is imported from the top level ``__init__.py``.
 
 For user projects wishing to patch the backend that they're using, this code
 should live in an :meth:`AppConfig.ready()<django.apps.AppConfig.ready>` method.
+
+Using operators to combine expressions
+--------------------------------------
+
+.. versionadded:: 2.0
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: I don't know if this might need some clarification. It might give the idea that CombinedExpression is new which isn't the case.

**Comment by timgraham** [comment]
@mrzechonek, did this drop off your radar?

**Comment by auvipy** [comment]
from another look, this pr doesn't seem to need a lot of work. clean up and some adjustment. only one concern could be using overload instead of multiple dispatch....

**Comment by auvipy** [comment]
@felixxm do you have interest on this issue? if not I would like to clean this up in a new PR.

**Comment by felixxm** [comment]
I wasn't involve in this PR. Feel-free to work on it.

**Comment by auvipy** [comment]
okk

**Comment by LuRsT** [comment]
Shall we close this PR since we got  #11310 now? @timgraham

**Comment by felixxm** [comment]
Superseded by #11310.

**Comment by mrzechonek** [comment]
:wave:
